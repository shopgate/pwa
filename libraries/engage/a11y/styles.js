import { css } from 'glamor';

// global class for screen readers only, visually hidden
css.global('.sr-only', {
  border: '0',
  clip: 'rect(0 0 0 0)',
  height: '1px',
  margin: '-1px',
  overflow: 'hidden',
  padding: '0',
  position: 'absolute',
  width: '1px',
  whiteSpace: 'nowrap',
  top: 0,
  left: 0,
});
