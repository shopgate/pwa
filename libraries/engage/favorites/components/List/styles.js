import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { colors, variables } = themeConfig;

const styles = {
  root: css({
    margin: '8px 8px 10px',
  }).toString(),
  rootNoFavoritesLists: css({
    background: colors.light,
    flexGrow: 1,
    paddingTop: variables.gap.xsmall,
    paddingLeft: '16px',
    paddingRight: '16px',
  }),
  title: css({
    flex: 1,
  }),
  divider: css({
    height: 1,
    width: 'calc(100% + 32px)',
    backgroundColor: 'rgb(234, 234, 234)',
    marginLeft: -16,
    marginRight: -16,
    marginBottom: 16,
  }),
  loadMoreButton: css({
    width: 'calc(100% - 32px)',
    margin: '16px 16px 0 16px',
    borderRadius: 5,
  }).toString(),
};

export default styles;
