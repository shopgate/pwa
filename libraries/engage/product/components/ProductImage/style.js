import { css } from 'glamor';
import { themeShadows, themeColors } from '@shopgate/pwa-common/helpers/config';

/**
 * Returns the correct styling for the given ratio
 * @param {number} imageRatio imageRatio
 * @return {string}
 */
const placeholderContainer = imageRatio => css({
  position: 'relative',
  width: '100%',
  ':before': {
    display: 'block',
    content: '""',
    width: '100%',
    paddingTop: `${100 * imageRatio}%`,
  },
}).toString();

const placeholderContent = css({
  position: 'absolute',
  width: '100%',
  height: '100%',
  top: 0,
  textAlign: 'center',
}).toString();

const placeholderIconScale = 0.65;

const placeholder = css({
  position: 'absolute',
  width: `${placeholderIconScale * 100}% !important`,
  height: `${placeholderIconScale * 100}% !important`,
  top: `${(1.0 - placeholderIconScale) * 50}%`,
  left: `${(1.0 - placeholderIconScale) * 50}%`,
  color: themeColors.placeholder,
}).toString();

const innerShadow = css({
  position: 'relative',
  overflow: 'hidden',
  ':after': {
    display: 'block',
    content: '""',
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    boxShadow: themeShadows.productImage,
    pointerEvents: 'none',
  },
}).toString();

export default {
  placeholderContainer,
  placeholderContent,
  placeholder,
  innerShadow,
};
