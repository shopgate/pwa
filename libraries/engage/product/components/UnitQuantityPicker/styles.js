import { css } from 'glamor';

export const small = css({
  width: 120,
}).toString();

export const big = css({
  width: 170,
}).toString();
