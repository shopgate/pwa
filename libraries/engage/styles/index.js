export * from './helpers';
export {
  ThemeProvider,
  createTheme,
  useTheme,
} from './theme';
export {
  makeStyles,
  withStyles,
  tss,
  GlobalStyles,
} from './tss';
