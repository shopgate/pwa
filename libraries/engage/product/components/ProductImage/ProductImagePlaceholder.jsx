import React, { useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import PlaceholderIcon from '@shopgate/pwa-ui-shared/icons/PlaceholderIcon';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';
import { useLoadImage } from '@shopgate/engage/core';
import styles from './style';

const { colors } = themeConfig;

/**
 * The ProductImagePlaceholder component.
 * @param {Object} props The components props.
 * @returns {JSX}
 */
const ProductImagePlaceholder = ({
  src, showInnerShadow, noBackground,
}) => {
  const [showPlaceholder, setShowPlaceholder] = useState(true);

  const srcLoaded = useLoadImage(src);
  useEffect(() => { setShowPlaceholder(false); }, [srcLoaded]);

  const contentStyles = useMemo(() => {
    if (srcLoaded) {
      return {
        backgroundImage: `url(${src})`,
        backgroundSize: 'contain',
        backgroundRepeat: 'no-repeat',
        backgroundPosition: 'center',
        backgroundColor: noBackground ? 'transparent' : colors.light,
        position: 'absolute',
      };
    }
    return {
      backgroundColor: noBackground ? 'transparent' : colors.light,
    };
  }, [src, noBackground, srcLoaded]);

  const contentClasses = classnames(styles.placeholderContent, {
    [styles.innerShadow]: showInnerShadow,
  });

  return (
    <div className={contentClasses} style={contentStyles} data-test-id="placeHolder">
      { showPlaceholder && <PlaceholderIcon className={styles.placeholder} />}
      { !showPlaceholder && ' ' }
    </div>
  );
};

ProductImagePlaceholder.propTypes = {
  noBackground: PropTypes.bool,
  showInnerShadow: PropTypes.bool,
  src: PropTypes.string,
};

ProductImagePlaceholder.defaultProps = {
  src: null,
  noBackground: false,
  showInnerShadow: false,
};

export default ProductImagePlaceholder;
