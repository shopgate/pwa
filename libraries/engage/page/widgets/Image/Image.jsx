import React from 'react';
import { makeStyles } from '@shopgate/engage/styles';
import { Link, ConditionalWrapper } from '@shopgate/engage/components';
import { ResponsiveWidgetImage } from '@shopgate/engage/page/components';
import { useImageWidget } from './hooks';

const useStyles = makeStyles()(() => ({
  link: {},
  image: {
    width: '100%',
  },
}));
/**
 * The ImageWidget is used to display a image.
 * @returns {JSX.Element}
 */
const Image = () => {
  const { cx, classes } = useStyles();
  const { link, altText, url } = useImageWidget();

  if (!url) return null;

  return (
    <ConditionalWrapper
      condition={!!link}
      wrapper={children => (
        <Link href={link} className={cx(classes.link)}>
          {children}
        </Link>
      )}
    >
      <ResponsiveWidgetImage
        src={url}
        alt={altText}
        className={cx(classes.image)}
      />
    </ConditionalWrapper>
  );
};
export default Image;
