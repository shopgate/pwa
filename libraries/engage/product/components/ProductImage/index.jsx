import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import classnames from 'classnames';
import { logger } from '@shopgate/pwa-core';
import appConfig, { themeConfig } from '@shopgate/pwa-common/helpers/config';
import Image from '@shopgate/pwa-common/components/Image';
import PlaceholderIcon from '@shopgate/pwa-ui-shared/icons/PlaceholderIcon';
import SurroundPortals from '@shopgate/pwa-common/components/SurroundPortals';
import { withWidgetSettings } from '../../../core/hocs/withWidgetSettings';
import { PORTAL_PRODUCT_IMAGE } from '../../../components/constants';
import ProductImagePlaceholder from './ProductImagePlaceholder';
import styles from './style';
import connect from './connector';

const { colors } = themeConfig;

/**
 * The product image component.
 * This component will behave like the core Image component with the additional
 * feature of showing a placeholder in case no src property has been passed
 * or the given source image cannot be loaded.
 */
class ProductImage extends Component {
  /**
   * See Image component manual for detailed description about the component property types.
   */
  static propTypes = {
    alt: PropTypes.string,
    animating: PropTypes.bool,
    'aria-hidden': PropTypes.bool,
    className: PropTypes.string,
    forcePlaceholder: PropTypes.bool,
    highestResolutionLoaded: PropTypes.func,
    noBackground: PropTypes.bool,
    placeholderSrc: PropTypes.string,
    ratio: PropTypes.arrayOf(PropTypes.number),
    resolutions: PropTypes.arrayOf(PropTypes.shape({
      width: PropTypes.number.isRequired,
      height: PropTypes.number.isRequired,
      blur: PropTypes.number,
    })),
    src: PropTypes.string,
    srcmap: PropTypes.arrayOf(PropTypes.string),
    widgetSettings: PropTypes.shape(),
  };

  static defaultProps = {
    alt: null,
    animating: true,
    'aria-hidden': null,
    className: null,
    forcePlaceholder: false,
    highestResolutionLoaded: () => { },
    noBackground: false,
    ratio: null,
    resolutions: [
      {
        width: 50,
        height: 50,
        blur: 2,
      },
      {
        width: 440,
        height: 440,
      },
    ],
    src: null,
    srcmap: null,
    placeholderSrc: null,
    widgetSettings: {},
  };

  /**
   * Component constructor
   * @param {Object} props The component properties
   */
  constructor(props) {
    super(props);
    logger.assert(!props.srcmap, 'Use of srcmap prop is deprecated. Use resolutions instead');

    const showPlaceholder = !props.src && (props.srcmap === null || props.srcmap.length === 0);
    this.state = {
      showPlaceholder,
    };
  }

  /**
   * Called when the component props change.
   * @param {Object} nextProps The new component properties
   */
  UNSAFE_componentWillReceiveProps(nextProps) {
    // Disable the placeholder to give the real image a new chance to load.
    // If we do not have a src property set then just show the placeholder instead.
    const showPlaceholder = !nextProps.src && (!nextProps.srcmap || nextProps.srcmap.length === 0);
    this.setState({
      showPlaceholder,
    });
  }

  /**
   * Should component update given the new props?
   * @param {Object} nextProps The next component props.
   * @param {Object} nextState The next state.
   * @return {boolean} Update or not.
   */
  shouldComponentUpdate(nextProps, nextState) {
    return !isEqual(this.props, nextProps) || !isEqual(this.state, nextState);
  }

  /**
   * Triggered when the image could not be loaded for some reason.
   */
  imageLoadingFailed = () => {
    this.setState({
      showPlaceholder: true,
    });
  };

  /**
   * Sets the image ratio based on width and height.
   * @return {number} The image ratio.
   */
  getImageRatio = () => {
    if (this.props.ratio) {
      const [x, y] = this.props.ratio;
      return ((y / x)).toFixed(3);
    }

    const { width, height } = this.props.resolutions[this.props.resolutions.length - 1];
    return ((height / width)).toFixed(3);
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      noBackground, className, placeholderSrc, alt,
    } = this.props;
    let { showInnerShadow } = this.props.widgetSettings;

    if (typeof showInnerShadow === 'undefined') {
      showInnerShadow = !appConfig.hideProductImageShadow;
    }

    if (this.state.showPlaceholder) {
      // Image is not present or could not be loaded, show a placeholder.
      return (
        <SurroundPortals portalName={PORTAL_PRODUCT_IMAGE}>
          <div
            className={classnames(styles.placeholderContainer(this.getImageRatio()), {
              [styles.innerShadow]: showInnerShadow,
              [className]: !!className,
            })}
            aria-hidden={this.props['aria-hidden']}
          >
            { placeholderSrc ? (
              <ProductImagePlaceholder
                src={placeholderSrc}
                alt={alt}
                showInnerShadow={showInnerShadow}
                noBackground={noBackground}
              />
            ) : (
              <div className={styles.placeholderContent} data-test-id="placeHolder">
                <PlaceholderIcon className={styles.placeholder} />
              </div>
            )}

          </div>
        </SurroundPortals>
      );
    }

    // Return the actual image.
    return (
      <SurroundPortals portalName={PORTAL_PRODUCT_IMAGE}>
        <div aria-hidden={this.props['aria-hidden']} className={className}>
          <Image
            {...this.props}
            className={showInnerShadow ? styles.innerShadow : ''}
            backgroundColor={noBackground ? 'transparent' : colors.light}
            onError={this.imageLoadingFailed}
          />
        </div>
      </SurroundPortals>
    );
  }
}

export { ProductImage as UnwrappedProductImage };

export default connect(withWidgetSettings(ProductImage, '@shopgate/engage/product/ProductImage'));
