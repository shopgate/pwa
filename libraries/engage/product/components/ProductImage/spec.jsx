import React from 'react';
import { shallow } from 'enzyme';
import Image from '@shopgate/pwa-common/components/Image';
import appConfig from '@shopgate/pwa-common/helpers/config';
import PlaceholderIcon from '@shopgate/pwa-ui-shared/icons/PlaceholderIcon';
import styles from './style';
import ProductImage from './index';

jest.unmock('@shopgate/pwa-core');
jest.mock('../../../core/hocs/withWidgetSettings');
jest.mock('@shopgate/pwa-common/helpers/config');
jest.mock('./connector', () => Component => Component);
jest.mock('@shopgate/engage/components', () => ({
  Image: () => null,
}));

describe('<ProductImage />', () => {
  it('should render a placeholder if no src prop is provided', () => {
    const wrapper = shallow(<ProductImage />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Image).length).toBe(0);
    expect(wrapper.find(PlaceholderIcon).length).toBe(1);
  });

  it('should render the image without a placeholder', () => {
    const wrapper = shallow(<ProductImage src="http://placehold.it/300x300" />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Image).length).toBe(1);
    expect(wrapper.find(PlaceholderIcon).length).toBe(0);
  });

  it('should not apply an inner shadow to the placeholder if turned off via the app config', () => {
    jest.spyOn(appConfig, 'hideProductImageShadow', 'get').mockReturnValue(true);
    const wrapper = shallow(<ProductImage />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(0);
  });

  it('should not apply an inner shadow to the image if turned off via the app config', () => {
    jest.spyOn(appConfig, 'hideProductImageShadow', 'get').mockReturnValue(true);
    const wrapper = shallow(<ProductImage src="http://placehold.it/300x300" />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(0);
  });

  it('should not apply an inner shadow to the placeholder if turned off via the widget settings', () => {
    const wrapper = shallow(<ProductImage widgetSettings={{ showInnerShadow: false }} />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(0);
  });

  it('should not apply an inner shadow to the image if turned off via the widget settings', () => {
    const wrapper = shallow(<ProductImage src="http://placehold.it/300x300" widgetSettings={{ showInnerShadow: false }} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(0);
  });

  it('should apply an inner shadow to the placeholder if turned off via the widget settings', () => {
    const wrapper = shallow(<ProductImage widgetSettings={{ showInnerShadow: true }} />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(1);
  });

  it('should apply an inner shadow to the image if turned off via the widget settings', () => {
    const wrapper = shallow(<ProductImage src="http://placehold.it/300x300" widgetSettings={{ showInnerShadow: true }} />).dive();

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(`.${styles.innerShadow}`).length).toBe(1);
  });
});
