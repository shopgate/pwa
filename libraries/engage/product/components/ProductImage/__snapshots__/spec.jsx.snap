// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductImage /> should apply an inner shadow to the image if turned off via the widget settings 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className={null}
  >
    <Image
      alt={null}
      animating={true}
      aria-hidden={null}
      backgroundColor="#fff"
      className="css-3hjqnz"
      classNameImg=""
      forcePlaceholder={false}
      highestResolutionLoaded={[Function]}
      noBackground={false}
      onError={[Function]}
      onLoad={null}
      placeholderSrc={null}
      ratio={null}
      resolutions={
        Array [
          Object {
            "blur": 2,
            "height": 50,
            "width": 50,
          },
          Object {
            "height": 440,
            "width": 440,
          },
        ]
      }
      src="http://placehold.it/300x300"
      srcmap={null}
      transition={null}
      unwrapped={false}
      widgetSettings={
        Object {
          "showInnerShadow": true,
        }
      }
    />
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should apply an inner shadow to the placeholder if turned off via the widget settings 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className="css-y5kkpq css-3hjqnz"
  >
    <div
      className="css-1vq3byi"
      data-test-id="placeHolder"
    >
      <Placeholder
        className="css-1g8mrwn"
      />
    </div>
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should not apply an inner shadow to the image if turned off via the app config 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className={null}
  >
    <Image
      alt={null}
      animating={true}
      aria-hidden={null}
      backgroundColor="#fff"
      className=""
      classNameImg=""
      forcePlaceholder={false}
      highestResolutionLoaded={[Function]}
      noBackground={false}
      onError={[Function]}
      onLoad={null}
      placeholderSrc={null}
      ratio={null}
      resolutions={
        Array [
          Object {
            "blur": 2,
            "height": 50,
            "width": 50,
          },
          Object {
            "height": 440,
            "width": 440,
          },
        ]
      }
      src="http://placehold.it/300x300"
      srcmap={null}
      transition={null}
      unwrapped={false}
      widgetSettings={
        Object {
          "mocked": "widgetSetting",
        }
      }
    />
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should not apply an inner shadow to the image if turned off via the widget settings 1`] = `
<ProductImage
  alt={null}
  animating={true}
  aria-hidden={null}
  className={null}
  forcePlaceholder={false}
  highestResolutionLoaded={[Function]}
  noBackground={false}
  placeholderSrc={null}
  ratio={null}
  resolutions={
    Array [
      Object {
        "blur": 2,
        "height": 50,
        "width": 50,
      },
      Object {
        "height": 440,
        "width": 440,
      },
    ]
  }
  src="http://placehold.it/300x300"
  srcmap={null}
  widgetSettings={
    Object {
      "showInnerShadow": false,
    }
  }
/>
`;

exports[`<ProductImage /> should not apply an inner shadow to the placeholder if turned off via the app config 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className="css-y5kkpq"
  >
    <div
      className="css-1vq3byi"
      data-test-id="placeHolder"
    >
      <Placeholder
        className="css-1g8mrwn"
      />
    </div>
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should not apply an inner shadow to the placeholder if turned off via the widget settings 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className="css-y5kkpq"
  >
    <div
      className="css-1vq3byi"
      data-test-id="placeHolder"
    >
      <Placeholder
        className="css-1g8mrwn"
      />
    </div>
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should render a placeholder if no src prop is provided 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className="css-y5kkpq"
  >
    <div
      className="css-1vq3byi"
      data-test-id="placeHolder"
    >
      <Placeholder
        className="css-1g8mrwn"
      />
    </div>
  </div>
</SurroundPortals>
`;

exports[`<ProductImage /> should render the image without a placeholder 1`] = `
<SurroundPortals
  portalName="component.product-image"
  portalProps={null}
>
  <div
    aria-hidden={null}
    className={null}
  >
    <Image
      alt={null}
      animating={true}
      aria-hidden={null}
      backgroundColor="#fff"
      className=""
      classNameImg=""
      forcePlaceholder={false}
      highestResolutionLoaded={[Function]}
      noBackground={false}
      onError={[Function]}
      onLoad={null}
      placeholderSrc={null}
      ratio={null}
      resolutions={
        Array [
          Object {
            "blur": 2,
            "height": 50,
            "width": 50,
          },
          Object {
            "height": 440,
            "width": 440,
          },
        ]
      }
      src="http://placehold.it/300x300"
      srcmap={null}
      transition={null}
      unwrapped={false}
      widgetSettings={
        Object {
          "mocked": "widgetSetting",
        }
      }
    />
  </div>
</SurroundPortals>
`;
